src/main/java/de/zalando/aruha/nakadi/service/subscription/ExactWeightRebalancer.java

```diff
@@ -0,0 +1,113 @@
+                candidates.remove(toTakeItem);
+                unassignedPartitions.add(toTakeItem);
+                toTake -= 1;
+            }
+        }
+
+        if (!unassignedPartitions.isEmpty()) {
+            // 3. Assign partitions to any nodes who are waiting for it.
+            final List<Partition> result = new ArrayList<>();
+            for (int idx = 0; idx < activeClientIds.size(); ++idx) {
+                final String clientId = activeClientIds.get(idx);
+
+                final int suggestedCount = partitionsPerClient[idx];
+                final int currentCount = partitions.containsKey(clientId) ? partitions.get(clientId).size() : 0;
+                for (int i = 0; i < suggestedCount - currentCount; ++i) {
+                    final Partition toMove = unassignedPartitions.iterator().next();
+                    unassignedPartitions.remove(toMove);
+
+                    if (toMove.getState() == Partition.State.UNASSIGNED) {
+                        result.add(toMove.toState(Partition.State.ASSIGNED, clientId, null));
+                    } else if (toMove.getState() == Partition.State.ASSIGNED) {
+                        if (toMove.getSession().equals(clientId)) {
+                            result.add(toMove.toState(Partition.State.ASSIGNED, clientId, null));
+                        } else {
+                            if (activeClientIds.contains(toMove.getSession())) {
+                                result.add(toMove.toState(Partition.State.REASSIGNING, toMove.getSession(), clientId));
+                            } else {
+                                result.add(toMove.toState(Partition.State.ASSIGNED, clientId, null));
+                            }
+                        }
+                    } else { // REASSIGNING
+                        if (activeClientIds.contains(toMove.getSession())) {
+                            // Taking from existing client.
+                            result.add(toMove.toState(Partition.State.REASSIGNING, toMove.getSession(), clientId));
+                        } else {
+                            // Client was dead, directly start streaming
+                            result.add(toMove.toState(Partition.State.ASSIGNED, clientId, null));
+                        }
+                    }
+                }
+            }
+            return result.toArray(new Partition[result.size()]);
+        } else {
+            return null;
+        }
+    }
+
+    static int[] splitByWeight(final int itemCount, final int[] weigths) {
+        final int totalWeight = IntStream.of(weigths).sum();
+        final int fixed = itemCount / totalWeight;
+        final int[] result = IntStream.of(weigths).map(w -> fixed * w).toArray();
+        if (fixed == 0) {
+            Arrays.fill(result, 1);
+        }
+
+        int left = itemCount - IntStream.of(result).sum();
+
+        // Yes, it's bad way, I know. But failed to figure out other simple way for exact case.
```

Comment:
@antban you're right. I could change it to behave as expected but then it would no longer be so simple. I missed some details there. So, just let it as is 👍 

```java
static int[] splitByWeight(final int itemCount, final int[] weigths) {
        final int totalWeight = IntStream.of(weigths).sum();
        int left = itemCount;
        int[] result = new int[weigths.length]

        for (int i = 0; i < weigths.length; ++i) {
             final float fixed = (float) itemCount / (float) totalWeight;
             final int v = (int) Math.ceil(Math.min(left, weigths[i] * fixed));
             result[i] = v;
             left -= v;
             itemCount -= v;
             totalWeight -= weigths[i];
         }
         return result;
}
```